libcob/profiling.c

```diff
@@ -0,0 +1,533 @@
+			module = info->procedures[0].text;
+			section = info->procedures[proc->section].text;
+			paragraph = proc->text;
+			entry = "";
+			kind = "PARAGRAPH";
+			break;
+
+		case COB_PROF_PROCEDURE_ENTRY:
+			module = info->procedures[0].text;
+			section = info->procedures[
+				info->procedures[proc->section].section].text;
+			paragraph = info->procedures[proc->section].text;
+			entry = proc->text;
+			kind = "ENTRY";
+			break;
+		case COB_PROF_PROCEDURE_CALL:
+			module = info->procedures[0].text;
+			section = info->procedures[
+				info->procedures[proc->section].section].text;
+			paragraph = info->procedures[proc->section].text;
+			entry = proc->text;
+			kind = "CALL";
+			break;
+		}
+	} else {
+		module = "program-id";
+		section = "section";
+		paragraph = "paragraph";
+		entry = "entry";
+		kind = "kind";
+		source_file = "file";
+		ncalls = 0;
+	}
+
+	for (i = 0; cobsetptr->cob_prof_format[i] != 0; i++) {
+		if (cobsetptr->cob_prof_format[i] == '%') {
+			i++;
+			switch (cobsetptr->cob_prof_format[i]) {
+			case 'M':
+			case 'm':
+				fputs (module, file);
+				break;
+			case 'S':
+			case 's':
+				fputs (section, file);
+				break;
+			case 'P':
+			case 'p':
+				fputs (paragraph, file);
+				break;
+			case 'E':
+			case 'e':
+				fputs (entry, file);
+				break;
+			case 'F':
+			case 'f':
+				fputs (source_file, file);
+				break;
+			case 'L':
+			case 'l':
+				if (info){
+					fprintf (file, "%d", line);
+				} else {
+					fputs ("line", file);
+				}
+				break;
+			case 'I':
+			case 'i':
+				if (info){
+					fprintf (file, "%d", cob_sys_getpid());
+				} else {
+					fputs ("pid", file);
+				}
+				break;
+			case 'W':
+			case 'w':
+				if (info){
+					fprintf (file, "%s:%d", source_file, line);
+				} else {
+					fputs ("location", file);
+				}
+				break;
+			case 'K':
+			case 'k':
+				fputs (kind, file);
+				break;
+			case 'T':
+			case 't':
+				if (info){
+					fprintf (file, "%lld", time);
+				} else {
+					fputs ("time-ns", file);
+				}
+				break;
+			case 'H':
+			case 'h':
+				if (info){
+					print_monotonic_time (file, time);
+				} else {
+					fputs ("time", file);
+				}
+				break;
+			case 'N':
+			case 'n':
+				if (info){
+					fprintf (file, "%d", ncalls);
+				} else {
+					fputs ("ncalls", file);
+				}
+				break;
+			default:
+				fputc ('%', file);
+				fputc (cobsetptr->cob_prof_format[i], file);
+			}
+		} else {
+			fputc (cobsetptr->cob_prof_format[i], file);
+		}
+	}
+	fputc ('\n', file);
+	fflush (file);
+}
+
+
+void
+cob_prof_end ()
+{
+	FILE *file;
+	const char* prof_filename = NULL;
+	struct cob_prof_module_list *l;
+
+	prof_init_static ();
+
+	if (!cobsetptr || !is_active || !prof_info_list) return;
+
+	while (current_idx >= 0) {
+		cob_prof_exit_procedure (called_runtimes[current_idx],
+					 called_procedures[current_idx]);
+	}
+
+	prof_filename = cobsetptr->cob_prof_filename;
+
+	file = fopen (prof_filename, !cobsetptr->cob_unix_lf ? "w" : "wb");
```

Comment:
This doesn't use the "common" approach. Please make `cob_open_logfile` an entry in coblocal.h and use it here; maybe replace `prof_filename` in the 3 cases by `cobsetptr->cob_prof_filename` directly